io/src/vlp32_grabber.cpp

```diff
@@ -0,0 +1,244 @@
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder(s) nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#include <pcl/io/vlp32_grabber.h>
+
+using boost::asio::ip::udp;
+
+/////////////////////////////////////////////////////////////////////////////
+pcl::VLP32Grabber::VLP32Grabber (const std::string& pcapFile) :
+    HDLGrabber ("", pcapFile)
+{
+    initializeLaserMapping ();
+    loadVLP32Corrections ();
+}
+
+/////////////////////////////////////////////////////////////////////////////
+pcl::VLP32Grabber::VLP32Grabber (const boost::asio::ip::address& ipAddress,
+                                 const std::uint16_t port) :
+    HDLGrabber (ipAddress, port)
+{
+    initializeLaserMapping ();
+    loadVLP32Corrections ();
+}
+
+/////////////////////////////////////////////////////////////////////////////
+pcl::VLP32Grabber::~VLP32Grabber () noexcept
+{
+}
+
+/////////////////////////////////////////////////////////////////////////////
+void
+pcl::VLP32Grabber::initializeLaserMapping ()
+{
+    for (std::uint8_t i = 0; i < VLP_MAX_NUM_LASERS / 2u; i++)
+    {
+        laser_rgb_mapping_[i * 2].b = static_cast<std::uint8_t> (i * 6 + 64);
+        laser_rgb_mapping_[i * 2 + 1].b = static_cast<std::uint8_t> ((i + 16) * 6 + 64);
+    }
+}
+
+/////////////////////////////////////////////////////////////////////////////
+void
+pcl::VLP32Grabber::loadVLP32Corrections ()
+{
+    double vlp32_vertical_corrections[] = { -25, -1, -1.667, -15.639, -11.31, 0, -0.667, -8.843, -7.254, 0.333, -0.333, -6.148, -5.333, 1.333, 0.667, -4,
+                                            -4.667, 1.667, 1, -3.667, -3.333, 3.333, 2.333, -2.667, -3, 7, 4.667, -2.333, -2, 15, 10.333, -1.333 };
+    double vlp32_azimuth_corrections[] = {-1.4, 4.2, -1.4, 1.4, -1.4, 1.4, -4.2, 1.4, -1.4, 4.2, -1.4, 1.4, -4.2, 1.4, -4.2, 1.4,
+                                          -1.4, 4.2, -1.4, 4.2, -4.2, 1.4, -1.4, 1.4, -1.4, 1.4, -1.4, 4.2, -4.2, 1.4, -1.4, 1.4};
```

Comment:
How did you generate these magic numbers?

Reply:
Thanks for checking

So they are not magic number, it's the angular correction provided by Velodyne (https://velodynelidar.com/wp-content/uploads/2019/08/VLP-32c.zip)

You'd find similar correction value in HDLGrabber and the existing VLPGrabber. (Different model, different values, naturally).

Reply:
Hi, does that answer your question?